### PR TITLE
1.17 Chunk Support Rewrite

### PR DESCRIPTION
**Pull request by nickelpro**
This is not good code. I would venture to say this is terrible code.

I Ctrl-C + Ctrl-V'd 1.16 and unga-bunga'd my way until the tests passed, this is the result.

In my defense, all of pr-chunk is janky as hell

### CODE REVIEW

**Comment by rom1504** [comment]
congrats on making this work!
can you check if the API is the same in 1.16 and 1.17 ?
if not, can you check if it's possible to make it the same ? (even if some param should be passed as undefined in some version or similar ideas)
if really not possible, then the api doc https://github.com/PrismarineJS/prismarine-chunk#api should be adapted

File: src/pc/1.14/ChunkColumn.js [hunk 1]
```diff
@@ -197,6 +199,14 @@ module.exports = (Block, mcData) => {
       return this.sectionMask
     }
 
+    getMaskArray () {
```
Comment (rom1504) [inline src/pc/1.14/ChunkColumn.js:202]: is this new method useful ?

File: src/pc/1.14/ChunkColumn.js [hunk 2]
```diff
@@ -197,6 +199,14 @@ module.exports = (Block, mcData) => {
...
+      return [this.getMask()]
+    }
+
+    countVerticalSections () {
```
Comment (rom1504) [inline src/pc/1.14/ChunkColumn.js:206]: this should be added to api doc if it's useful

File: src/pc/1.14/ChunkColumn.js (hunk 1 above)
Comment (nickelpro) [reply src/pc/1.14/ChunkColumn.js:202]: No, it's a hold over from when I was rebasing from Archengius. Will remove

Comment (nickelpro) [reply src/pc/1.14/ChunkColumn.js:202]: Not new, present in 1.16

Comment (nickelpro) [reply src/pc/1.14/ChunkColumn.js:202]: I don't think it's useful either, but again I just Ctrl+C, Ctrl-V

File: src/pc/1.14/ChunkColumn.js (hunk 2 above)
Comment (nickelpro) [reply src/pc/1.14/ChunkColumn.js:206]: Again I would much rather just remove these methods, but they were present in 1.16 so for API consistency they're present here

File: src/pc/1.14/ChunkColumn.js (hunk 1 above)
Comment (rom1504) [reply src/pc/1.14/ChunkColumn.js:202]: https://github.com/PrismarineJS/prismarine-chunk/pull/133/files pretty sure it was added in this pr in all versions

File: src/pc/1.14/ChunkColumn.js (hunk 2 above)
Comment (rom1504) [reply src/pc/1.14/ChunkColumn.js:206]: same answer as other method, it was added in this pr

**Comment by nickelpro** [comment]
API is consistent with 1.16 with the exception of:
* Bitmaps/masks are now long arrays
* Removed `minWorldHeight` and `maxWorldHeight` since they're not useful or meaningful in 1.17

**Comment by nickelpro** [comment]
Now the only API inconsistency is the bitmap Number -> Array of protodef Longs, which I don't think is reducible complexity

**Review by Karang — APPROVED** [review]
LGTM

**Comment by rom1504** [comment]
new exposed things and changes to API should be documented  in the readme

**Review by rom1504 — APPROVED** [review]

**Pull request merged**